### PR TITLE
Feature - Login screen

### PR DESCRIPTION
**Pull request by sjacobs1**
##Description
Added login view model to handle login, and updated login UI

##Screenshots
 ! ! ! ! Login image is placeholder ! ! ! !
 
Screenshot 2024-05-06 at 9 04 12 AM

Screenshot 2024-05-06 at 8 19 34 AM

## Checklist 
-  The code follows the project's coding standards. 
- [ ] 100% Unit test coverage for View Models 
-  All warnings Resolved.

### CODE REVIEW

**Review by giacatano — DISMISSED** [review]

**Review by ksookdhew — COMMENTED** [review]
I would suggest putting the username and password labels in a darker colour to make them ore readable

**Review by ksookdhew — DISMISSED** [review]

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift [hunk 1]
```diff
@@ -1,68 +1,68 @@
+////
+////  ViewController.swift
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:2]: HomeScreenViewController

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift [hunk 2]
```diff
@@ -1,68 +1,68 @@
...
-            cell.posterCover.sd_setImage(with: posterURL, placeholderImage: UIImage(named: "placeholder"))
-        } else {
-            cell.posterCover.image = UIImage(named: "placeholder")
-        }
-
-        return cell
-    }
-
-}
+//    @IBOutlet weak var nowPlayingCollectionView: UICollectionView!
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:13]: is this whole file meant to be commented out?

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
+//
+//  Login-ViewModel.swift
+//  MovieAppPetProject
+//
+//  Created by Sebastian Jacobs on 2024/05/05.
+//
+
+// LoginViewModel.swift
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: can remove

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+import Foundation
+
+class LoginViewModel {
+
+    // MARK: - Variables
+    private let username = "Admin"
+    private let password = "DVTPassword"
+
+    // MARK: - Function
+    func login(username: String?, password: String?, completion: @escaping (Bool) -> Void) {
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: no need for a completion. just let the function return true or false

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        guard let username = username, !username.isEmpty,
+              let password = password, !password.isEmpty else {
+            completion(false)
+            return
+        }
+
+        let isValidCredentials = username == self.username && password == self.password
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: return username == self.username && password == self.password
when completion is removed

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 6]
```diff
@@ -9,49 +9,38 @@ import UIKit
...
 
-        let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
-        self.present(alert, animated: true)
-        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
-            alert.dismiss(animated: true)
+    @IBAction private func loginTapped(_ sender: Any) {
+        let username = usernameInput.text
+        let password = passwordInput.text
+
+        loginViewModel.login(username: username, password: password) { [weak self] success in
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: [weak self] success in can be removed when remove closure

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 7]
```diff
@@ -9,49 +9,38 @@ import UIKit
...
+            guard let self = self else { return }
+            if success {
+                self.performSegue(withIdentifier: "showHomeScreen", sender: self)
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: can add segue name to constants file

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift (hunk 2 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:13]: it is an old version of my Home Screen view controller, will remove it

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift (hunk 1 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:2]: will remove

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift (hunk 3 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: will do

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift (hunk 4 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: understood

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift (hunk 5 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: removed

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift (hunk 6 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: removed

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift (hunk 7 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: will do

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift [hunk 8]
```diff
@@ -1,68 +1,68 @@
...
+//        layout.scrollDirection = .horizontal
+//        layout.minimumInteritemSpacing = 10
+//        layout.minimumLineSpacing = 10
+////        layout.sectionInset = UIEdgeInsets(top: 5, left: 5, bottom: 20, right: 10)
+//        nowPlayingCollectionView.collectionViewLayout = layout
+//
+//        fetchNowPlayingMovies()
+//    }
+//
+//    func fetchNowPlayingMovies() {
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:34]: Rather delete the file than commenting out

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift [hunk 9]
```diff
@@ -0,0 +1,24 @@
+//
+//  Login-ViewModel.swift
+//  MovieAppPetProject
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: Please rename file name to LoginViewModel

File: MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift (hunk 8 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/HomeScreen/HomeScreen-ViewController.swift:34]: I did delete it 💪 not sure if its showing now

File: MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift (hunk 9 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/Login-ViewModel.swift]: will do

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 10]
```diff
@@ -9,49 +9,35 @@ import UIKit
...
-        let alert = UIAlertController(title: alertTitle, message: alertMessage, preferredStyle: .alert)
-        self.present(alert, animated: true)
-        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
-            alert.dismiss(animated: true)
-        }
+    @IBAction private func loginTapped(_ sender: Any) {
+        let username = usernameInput.text
+        let password = passwordInput.text
 
+        if loginViewModel.login(username: username, password: password) {
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: This logic should be done in the viewModel
Use Delegate methods

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 11]
```diff
@@ -9,49 +9,35 @@ import UIKit
...
-    @IBOutlet weak var passwordInput: UITextField!
+    // MARK: - Variable
+    private let loginViewModel = LoginViewModel()
 
+    // MARK: - Functions
     override func viewDidLoad() {
         super.viewDidLoad()
-
-        // Do any additional setup after loading the view.
+        usernameInput.attributedPlaceholder = NSAttributedString(string: "Username", attributes: [NSAttributedString.Key.foregroundColor: UIColor.gray])
```
Comment (kieran-woodrow) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift:23]: do you have to use NSAttribute?
1)Can you not do this in storyboard?
2)If not, can you not just do 

usernameInput.text = "Test..."
usernameInput.textColor = UIColor.gray
and same for password

**Review by kieran-woodrow — CHANGES_REQUESTED** [review]

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift (hunk 11 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift:23]: I tried on storyboard and had a cll with Time we it isn't working in there. will refactor to your suggestion 💪

Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift:23]: there doesn't seem to be any methods for colour change on placeholder, and textColour changes only input text colour

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 12]
```diff
@@ -5,53 +5,49 @@
 //  Created by Sebastian Jacobs on 2024/04/15.
 //
 
+protocol LoginViewModelDelegate: AnyObject {
+    func navigateToHomeScreen()
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: Move this declaration to ViewModel

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift [hunk 13]
```diff
@@ -5,53 +5,49 @@
...
+    func displayErros(with message: String)
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]:
```suggestion
    func displayError(message: String)
```

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift (hunk 12 above)
Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewController.swift]: will do

File: MovieAppPetProject/MovieAppPetProject/Login/LoginViewModel.swift [hunk 14]
```diff
@@ -5,6 +5,11 @@
 //  Created by Sebastian Jacobs on 2024/05/05.
 //
 
+protocol LoginViewModelDelegate: AnyObject {
+    func navigateToHomeScreen()
+    func displayErros(message: String)
```
Comment (simekanidev) [inline MovieAppPetProject/MovieAppPetProject/Login/LoginViewModel.swift]: displayError

Comment (sjacobs1) [reply MovieAppPetProject/MovieAppPetProject/Login/LoginViewModel.swift]: ah damn will change

**Review by simekanidev — APPROVED** [review]

**Review by giacatano — APPROVED** [review]

**Review by kieran-woodrow — APPROVED** [review]

**Pull request merged**